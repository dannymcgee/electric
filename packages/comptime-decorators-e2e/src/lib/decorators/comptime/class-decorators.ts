import { ClassDecorator } from "@electric/comptime-decorators";

export const ClassDecoratorFn: ClassDecorator = function (node, $) {
	return this.factory.updateClassDeclaration(
		node,
		node.modifiers,
		node.name,
		node.typeParameters,
		node.heritageClauses,
		node.members.concat(
			$.PropertyDecl(
				[],
				$.Identifier("test"),
				undefined,
				undefined,
				$.StringLiteral("Hello, world!")
			)
		)
	)
}

export function ClassDecoratorFactory(
	testString: string,
	testTemplate: string,
	testNumber: number,
	testBigint: bigint,
	testRegex: RegExp,
	testBool: boolean,
	testNil: null,
): ClassDecorator {
	return function (node, $) {
		return this.factory.updateClassDeclaration(
			node,
			node.modifiers,
			node.name,
			node.typeParameters,
			node.heritageClauses,
			node.members.concat(
				$.PropertyDecl(
					[],
					$.Identifier("testString"),
					undefined,
					undefined,
					$.StringLiteral(testString)
				),
				$.PropertyDecl(
					[],
					$.Identifier("testTemplate"),
					undefined,
					undefined,
					$.NoSubstitutionTemplateLiteral(testTemplate)
				),
				$.PropertyDecl(
					[],
					$.Identifier("testNumber"),
					undefined,
					undefined,
					$.NumericLiteral(testNumber)
				),
				$.PropertyDecl(
					[],
					$.Identifier("testBigint"),
					undefined,
					undefined,
					$.BigIntLiteral(`${testBigint.toString(10)}n`)
				),
				$.PropertyDecl(
					[],
					$.Identifier("testRegex"),
					undefined,
					undefined,
					$.RegularExpressionLiteral(testRegex.toString())
				),
				$.PropertyDecl(
					[],
					$.Identifier("testBool"),
					undefined,
					undefined,
					testBool ? $.True() : $.False()
				),
				$.PropertyDecl(
					[],
					$.Identifier("testNil"),
					undefined,
					undefined,
					testNil === null ? $.Null() : $.Identifier("undefined")
				),
			)
		)
	}
}
