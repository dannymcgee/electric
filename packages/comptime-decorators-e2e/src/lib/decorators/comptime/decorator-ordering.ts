import { PropertyDecorator } from "@electric/comptime-decorators";
import * as ts from "typescript";

export const first: PropertyDecorator = function (node, $) {
	return [
		$.updatePropertyDecl(
			node,
			node.modifiers,
			$.Identifier(`${(node.name as ts.Identifier).text}_first`),
			node.questionToken ?? node.exclamationToken,
			node.type,
			node.initializer
		),
		$.PropertyDecl(
			undefined,
			$.PrivateIdentifier("#firstAddtl"),
			undefined,
			undefined,
			$.StringLiteral("first")
		)
	]
}

export const second: PropertyDecorator = function (node, $) {
	return [
		$.updatePropertyDecl(
			node,
			node.modifiers,
			$.Identifier(`${(node.name as ts.Identifier).text}_second`),
			node.questionToken ?? node.exclamationToken,
			node.type,
			node.initializer
		),
		$.PropertyDecl(
			undefined,
			$.PrivateIdentifier("#secondAddtl"),
			undefined,
			undefined,
			$.StringLiteral("second")
		)
	]
}

export const third: PropertyDecorator = function (node, $) {
	return [
		$.updatePropertyDecl(
			node,
			node.modifiers,
			$.Identifier(`${(node.name as ts.Identifier).text}_third`),
			node.questionToken ?? node.exclamationToken,
			node.type,
			node.initializer
		),
		$.PropertyDecl(
			undefined,
			$.PrivateIdentifier("#thirdAddtl"),
			undefined,
			undefined,
			$.StringLiteral("third")
		)
	]
}
