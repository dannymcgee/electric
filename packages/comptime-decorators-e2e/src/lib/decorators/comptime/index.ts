export * from "./class-decorators"
export * from "./decorator-ordering"
export * from "./method-decorators"
export * from "./property-decorators"
