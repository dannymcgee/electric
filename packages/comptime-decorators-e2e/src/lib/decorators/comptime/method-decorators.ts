import { MethodDecorator } from "@electric/comptime-decorators";

export const methodDecoratorFn: MethodDecorator = function (node, $) {
	return this.factory.updateMethodDeclaration(
		node,
		node.modifiers,
		node.asteriskToken,
		node.name,
		node.questionToken,
		node.typeParameters,
		node.parameters,
		node.type,
		$.Block([
			$.ExpressionStmt(
				$.CallExpr(
					$.PropertyAccessExpr(
						$.Identifier("console"),
						$.Identifier("log")
					),
					undefined,
					[
						$.StringLiteral("Hello, world!")
					]
				)
			),
			...(node.body?.statements ?? [])
		])
	)
}

export function methodDecoratorFactory(): MethodDecorator {
	return methodDecoratorFn
}
