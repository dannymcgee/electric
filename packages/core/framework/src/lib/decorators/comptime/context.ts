import {
	PluginContext,
	ClassDecorator as _ClassDecorator,
	MethodDecorator as _MethodDecorator,
	PropertyDecorator as _PropertyDecorator,
} from "@electric/comptime-decorators"

import * as ts from "typescript"

export type ClassDecorator = _ClassDecorator<Context | undefined>
export type MethodDecorator = _ClassDecorator<Context | undefined>
export type PropertyDecorator = _PropertyDecorator<Context | undefined>

export class Context {
	#inputs = new WeakMap<ts.Node, ts.PropertyDeclaration[]>()

	addInput(parent: ts.Node, node: ts.PropertyDeclaration) {
		let inputs: ts.PropertyDeclaration[] = []
		if (!this.#inputs.has(parent))
			this.#inputs.set(parent, inputs)
		else
			inputs = this.#inputs.get(parent)!

		inputs.push(node)
	}

	getInputs(parent: ts.Node): ts.PropertyDeclaration[] {
		return this.#inputs.get(parent) ?? []
	}
}

export function getContext(this: PluginContext<Context | undefined>): Context {
	return this.userContext ??= new Context()
}
