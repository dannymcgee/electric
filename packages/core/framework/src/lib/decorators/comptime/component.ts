import { ClassDecorator } from "@electric/comptime-decorators"
import * as ts from "typescript"

import { getContext } from "./context"

export function Component(tagName: string): ClassDecorator {
	return function (node, $) {
		const cx = getContext.call(this)
		const inputs = cx.getInputs(this.original(node))

		const observedAttributes = $.PropertyDecl(
			[$.Token(ts.SyntaxKind.StaticKeyword)],
			$.Identifier("observedAttributes"),
			undefined,
			undefined,
			$.ArrayLiteralExpr(
				inputs.map(p => $.StringLiteral((p.name as ts.Identifier).text))
			)
		)

		return [
			$.ImportDecl(
				undefined,
				$.ImportClause(
					false,
					$.Identifier("__$setupInputs"),
					undefined
				),
				$.StringLiteral("@electric/framework/src/lib/internal/inputs")
			),
			$.ImportDecl(
				undefined,
				$.ImportClause(
					false,
					undefined,
					$.NamedImports([
						$.ImportSpecifier(
							false,
							$.Identifier("STRING"),
							$.Identifier("__$STRING")
						)
					])
				),
				$.StringLiteral("@electric/framework/src/lib/internal/serde")
			),
			$.updateClassDecl(
				node,
				node.modifiers,
				node.name,
				node.typeParameters,
				node.heritageClauses,
				[
					observedAttributes,
					...node.members
				]
			),
			$.ExpressionStmt(
				$.CallExpr(
					$.PropertyAccessExpr(
						$.Identifier("customElements"),
						$.Identifier("define")
					),
					undefined,
					[
						$.StringLiteral(tagName),
						node.name!
					]
				)
			),
			$.ExpressionStmt(
				$.CallExpr(
					$.Identifier("__$setupInputs"),
					undefined,
					[
						node.name!,
						$.ObjectLiteralExpr(
							inputs.map(p => $.PropertyAssignment(
								p.name,
								$.ObjectLiteralExpr([
									$.PropertyAssignment(
										$.Identifier("serde"),
										$.Identifier("__$STRING"),
									),
									$.PropertyAssignment(
										$.Identifier("init"),
										p.initializer ?? $.Identifier("undefined")
									)
								])
							))
						)
					]
				)
			)
		]
	}
}
