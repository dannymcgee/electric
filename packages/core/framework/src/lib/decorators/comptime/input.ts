import { getContext, PropertyDecorator } from "./context"

export const input: PropertyDecorator
= function (node, $) {
	getContext.call(this).addInput(this.original(node).parent, node)

	return node
}
