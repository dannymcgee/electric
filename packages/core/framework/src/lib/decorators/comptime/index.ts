import comptimeDecorators from "@electric/comptime-decorators"

import { Component } from "./component"

// Plugin hook for Nx custom transformers loader
export const before = comptimeDecorators({ Component })
