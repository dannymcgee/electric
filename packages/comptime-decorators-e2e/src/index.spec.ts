import comptimeDecoators, { TsCompiler } from "@electric/comptime-decorators"
import * as path from "path"

import * as decorators from "./lib/decorators/comptime"

describe("comptime-decorators", () => {
	it("e2e", () => {
		const tsConfigPath = path.resolve(__dirname, "../tsconfig.lib.json")
		const tsc = new TsCompiler(tsConfigPath)
		// FIXME: TypeScript is not recognizing the decorator facories in this
		// module as members of the Decorator union
		// console.log("decorators:", decorators)
		const transformerFactory = comptimeDecoators(decorators as any)

		const { virtualFs, diagnostics } = tsc.transformAll(transformerFactory)

		for (let diag of diagnostics)
			tsc.reportDiagnostics(diag)

		Array.from(virtualFs.keys())
			.filter(filename => filename.includes("comptime-decorators-e2e"))
			.filter(filename => filename.endsWith("src/index.js"))
			.forEach(filename => {
				const text = virtualFs.get(filename)!
				// console.log(`// ${filename}\n\n${text}`)
				expect(text).toMatchSnapshot()
			})
	})
})
