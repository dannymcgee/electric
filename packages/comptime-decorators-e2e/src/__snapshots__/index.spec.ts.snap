// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsCompiler should work 1`] = `
"import { ClassDecoratorFactory, ClassDecoratorFn, first, methodDecoratorFactory, methodDecoratorFn, propDecoratorFactory, propDecoratorFn, second, third, } from \\"./lib/decorators/runtime\\";
// \`ClassDecoratorFn\` should add a \`test\` field to \`DecoratorFnTest\` initialized
// to a value of \`\\"Hello, world!\\"\`.
export class DecoratorFnTest {
    // \`propDecoratorFn\` should replace \`decoratedProp\` with \`get\`/\`set\`
    // accessors backed by a private field.
    #decoratedProp = \\"Hello, world!\\";
    get decoratedProp() { return this.#decoratedProp; }
    set decoratedProp(value) { this.#decoratedProp = value; }
    // \`methodDecoratorFn\` should add a console.log call to the body of the
    // decorated method.
    hello() { console.log(\\"Hello, world!\\"); }
    test = \\"Hello, world!\\";
}
// \`ClassDecoratorFactory\` should add each argument to \`DecoratorFactoryTest\` as
// a class field initialized to its passed value.
export class DecoratorFactoryTest {
    // \`propDecoratorFactory\` should replace \`decoratedProp\` with \`get\`/\`set\`
    // accessors backed by a private field.
    #decoratedProp = \\"Hello, world!\\";
    get decoratedProp() { return this.#decoratedProp; }
    set decoratedProp(value) { this.#decoratedProp = value; }
    // \`methodDecoratorFactory\` should add a console.log call to the body of the
    // decorated method.
    hello() { console.log(\\"Hello, world!\\"); }
    testString = \\"Hello, world!\\";
    testTemplate = \`Hello, world!\`;
    testNumber = 42;
    testBigint = 123456789123456789123456789123456789n;
    testRegex = /[_a-zA-Z][_a-zA-Z0-9]*/g;
    testBool = true;
    testNil = null;
}
export class DecoratorOrdering {
    // These decorators should be applied in the order implied by their names.
    // The end result should see \`decoratedProp\` renamed to
    // \`decoratedProp_first_second_third\`, followed by three new properties in
    // reverse order, \`#thirdAddtl\`, \`#secondAddtl\`, then \`#firstAddtl\`
    decoratedProp_first_second_third;
    #thirdAddtl = \\"third\\";
    #secondAddtl = \\"second\\";
    #firstAddtl = \\"first\\";
}
"
`;
